### DISCUSSION TITLE
Polynomial Evaluation

### DISCUSSION THREAD

**Discussion by RagonEbker** [sympy/sympy#25571] [Q&A]
Hello, 
I want to understand how Sympy evaluates polynomials. Does it use horners method?

**Comment by oscarbenjamin** [comment]
Can you show code to demonstrate what you mean by SymPy evaluating a polynomial?

**Reply by RagonEbker** [reply]
For example here:

```
deg = 40
p = Poly(random_poly(x,deg,0,10),x)


start_time = time.time()
p(100)
print("--- %s seconds ---" % (time.time() - start_time))


q = p.as_expr()
start_time = time.time()
q.subs(x,100)
print("--- %s seconds ---" % (time.time() - start_time))
```

outputs:
--- 0.00011587142944335938 seconds ---
--- 0.028937339782714844 seconds ---

So, evaluating a poly is a lot faster than evaluating the expression itself. Does that lie in the usage of Horners method?

**Reply by oscarbenjamin** [reply]
The main difference is just that Poly uses more efficient internal representations than SymPy's ordinary symbolic expressions. For Poly  the method is here:
https://github.com/sympy/sympy/blob/2f80497c7e83d889748df102e9a0c126fdbaeab5/sympy/polys/densetools.py#L288-L314
